lib/spectrum/item.rb

```diff
@@ -38,53 +38,76 @@ class Holding
 
     ETAS_START = 'Full text available,'
 
-    attr_reader :holding, :record, :barcode
+    attr_reader :holding, :id 
 
-    def initialize(data, record, barcode)
-      @holding = extract_holding(data, record, barcode) || barcode_not_found
-      @item    = data[record]
-      @record  = record
-      @barcode = barcode
-      @etas_ids = extract_etas_ids(@item)
+    def initialize(id:, holdings:, item:)
+      @item = item || barcode_not_found # single item from getHoldings.pl element
+      @holdings  = holdings  #single element from getHoldings.pl
+      @id = id #doc_id
+      @etas_ids = extract_etas_ids(@holdings)
     end
 
-    def self.for(request:, source:)
+    def self.for_get_this(request:, source:)
```

Comment:
I have followed the (probably anti-) pattern of overriding `self.new` to accomplish these kinds of things.  The logic can get a little messy, but can be delegated to the subclass to decide if the initialization arguments match their criteria, which at some level I like.  But it can make things harder to see upfront.  Do you have opinions on that arrangement?  Colin F. in the past had not liked my approach because it's calling ClassName.new but getting an OtherSubClass instance.

There's an example in `lib/spectrum/holding/action.rb`.

Reply:
The main reason I've been using "for" instead of overriding "new" is that that's what Sandi Metz used in the 99bottles book and I liked it. What's going on is clear: it's like "new" but a bit different. I won't do the double take of looking at "initialize" and not understanding why the params for that don't match "new". I won't get confused when I call ClassName.new and it returns an OtherSubClass instance, as you mentioned. Also, I can have variations like Item.for_get_this and Item.for_barcode and that make more sense at first glance than something like Item.new_get_this would.

 (Item.for_get_this is unintentionally funny, though. *sigh*)

Reply:
I don't remember if it was in the book, but there was also the fact that method namespaces are different from class namespaces so you can do something like:

```ruby
module Spectrum
  class Item
  end
  def Item(...)
    case ...
    when 'plain item'
      Spectrum::Item.new(...)
    when 'special item'
      Spectrum::SpecialItem.new(...)
    end
  end
end

Spectrum::Item(...)
```

I'm not saying any one way is better, just that you might see all three in the code depending on which I was thinking about when I wrote it.  There's probably an even more confusing variant, where I define `self.new` on a module and use that to instantiate a class under the module.